validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java

```diff
@@ -0,0 +1,150 @@
+              currentEpoch = currentEpoch.plus(UnsignedLong.ONE)) {
+            scheduleDutiesForEpoch(currentEpoch).reportExceptions();
+          }
+          return startEpoch.compareTo(endEpoch) > 0 ? startEpoch : endEpoch;
+        });
+  }
+
+  private SafeFuture<Void> scheduleDutiesForEpoch(final UnsignedLong epoch) {
+    LOG.trace("Requesting duties for epoch {}", epoch);
+    return validatorApiChannel
+        .getDuties(epoch, validators.keySet())
+        .orTimeout(Constants.VALIDATOR_DUTIES_TIMEOUT, TimeUnit.SECONDS)
+        .thenApply(
+            maybeDuties ->
+                maybeDuties.orElseThrow(
+                    () ->
+                        new IllegalStateException(
+                            "Duties could not be calculated because chain data was not yet available")))
+        .thenAccept(duties -> duties.forEach(this::scheduleDuties))
+        .exceptionallyCompose(
+            error -> {
+              LOG.error(
+                  "Failed to request validator duties for epoch "
+                      + epoch
+                      + ". Retrying after delay.",
+                  error);
+              return asyncRunner.runAfterDelay(
+                  () -> scheduleDutiesForEpoch(epoch), 5, TimeUnit.SECONDS);
+            });
+  }
+
+  private void scheduleDuties(final ValidatorDuties validatorDuties) {
+    LOG.trace("Got validator duties: {}", validatorDuties);
+    final Validator validator = validators.get(validatorDuties.getPublicKey());
+    validatorDuties
+        .getDuties()
+        .ifPresent(
+            duties -> {
+              duties
+                  .getBlockProposalSlots()
+                  .forEach(slot -> scheduleBlockProduction(validator, slot));
+              scheduleAttestationProduction(
+                  duties.getAttestationCommitteeIndex(),
+                  validator,
+                  duties.getValidatorIndex(),
+                  duties.getAttestationSlot());
+            });
+  }
+
+  private void scheduleBlockProduction(final Validator validator, final UnsignedLong slot) {
+    blockProposalDuties.put(slot, dutyFactory.createBlockProductionDuty(validator, slot));
+  }
+
+  private void scheduleAttestationProduction(
+      final int attestationCommitteeIndex,
+      final Validator validator,
+      final int validatorIndex,
+      final UnsignedLong slot) {
+    attestationProposalDuties
+        .computeIfAbsent(slot, dutyFactory::createAttestationProductionDuty)
+        .addValidator(attestationCommitteeIndex, validator, validatorIndex);
+  }
+
+  @Override
+  public void onBlockProductionDue(final UnsignedLong slot) {
+    final BlockProductionDuty duty = blockProposalDuties.remove(slot);
+    if (duty == null) {
+      return;
+    }
+    duty.performDuty();
+  }
+
+  @Override
+  public void onAttestationCreationDue(final UnsignedLong slot) {
+    final AttestationProductionDuty duty = attestationProposalDuties.remove(slot);
```

Comment:
nit: would be nice to have a generic function that takes attestationProposalDuties and blockProposalDuties as arguments since the rest of the logic is same.